test/framework/src/main/java/org/elasticsearch/test/transport/CapturingTransport.java

```diff
@@ -94,26 +95,31 @@ public CapturedRequest(DiscoveryNode node, long requestId, String action, Transp
      * @return the captured requests
      */
     public CapturedRequest[] getCapturedRequestsAndClear() {
-        CapturedRequest[] capturedRequests = capturedRequests();
-        clear();
-        return capturedRequests;
+        List<CapturedRequest> requests = new ArrayList<>(capturedRequests.size());
+        capturedRequests.drainTo(requests);
+        return requests.toArray(new CapturedRequest[0]);
+    }
+
+    private Map<String, List<CapturedRequest>> groupRequestsByTargetNode(Collection<CapturedRequest> requests) {
```

Comment:
I opened https://github.com/elastic/elasticsearch/pull/31312 for CapturingTransport .